app/buck2_events/src/sink/remote.rs

```diff
@@ -334,23 +334,618 @@ mod fbcode {
+                    // This means we can't have `$` in a header key or value, which isn't great. On the
+                    // flip side, env vars are good for things like credentials, which those headers
+                    // are likely to contain. In time, we should allow escaping.
+                    let key = substitute_env_vars(&h.key)?;
+                    let value = substitute_env_vars(&h.value)?;
+
+                    let key = MetadataKey::<metadata::Ascii>::from_bytes(key.as_bytes())
+                        .with_context(|| format!("Invalid key in header: `{}: {}`", key, value))?;
+
+                    let value = MetadataValue::try_from(&value)
+                        .with_context(|| format!("Invalid value in header: `{}: {}`", key, value))?;
+
+                    anyhow::Ok((key, value))
+                })
+                .collect::<Result<_, _>>()
+                .context("Error converting headers")?;
+
+            Ok(Self {
+                headers: Arc::new(headers),
+            })
+        }
+    }
+
+    impl Interceptor for InjectHeadersInterceptor {
+        fn call(
+            &mut self,
+            mut request: tonic::Request<()>,
+        ) -> Result<tonic::Request<()>, tonic::Status> {
+            for (k, v) in self.headers.iter() {
+                request.metadata_mut().insert(k.clone(), v.clone());
+            }
+            Ok(request)
+        }
+    }
+
+    type GrpcService = InterceptedService<Channel, InjectHeadersInterceptor>;
+
+    async fn connect_build_event_server() -> anyhow::Result<PublishBuildEventClient<GrpcService>> {
+        let uri = std::env::var("BES_URI")?.parse()?;
+        let mut channel = Channel::builder(uri);
+        let tls_config = ClientTlsConfig::new();
+        {
+            let tls_setting = std::env::var("BES_TLS").unwrap_or("0".to_owned());
+            match tls_setting.as_str() {
+                "1" | "true" => {
+                    channel = channel.tls_config(tls_config)?;
+                },
+                _ => {},
+            }
+        }
+        // TODO: parse PEM
+        let endpoint = channel
+            .connect()
+            .await
+            .context("connecting to Bazel event stream gRPC server")?;
+        let mut headers = vec![];
+        for hdr in std::env::var("BES_HEADERS").unwrap_or("".to_owned()).split(",") {
+            let hdr = hdr.trim();
+            if !hdr.is_empty() {
+                headers.push(HttpHeader::from_str(hdr)?);
+            }
+        };
+        let interceptor = InjectHeadersInterceptor::new(&headers)?;
+        let client = PublishBuildEventClient::with_interceptor(endpoint, interceptor);
+        Ok(client)
+    }
+
+    fn buck_to_bazel_events<S: Stream<Item = BuckEvent>>(events: S) -> impl Stream<Item = v1::BuildEvent> {
+        let mut target_actions: HashMap<(String, String), Vec<(BuildEventId, bool)>> = HashMap::new();
+        stream! {
+            for await event in events {
```

Comment:
you should be able to use unpack_event to get an UnpackedBuckEvent that cuts down a bit on the boilerplate here (compare superconsole's matching on these events: 

https://github.com/facebook/buck2/blob/86174c3e5fddc2bf0e583dc7ed39984bfd61891e/app/buck2_client_ctx/src/subscribers/superconsole.rs#L504